Recorder tests that use Recorder.block_till_done are unreliable

<!-- READ THIS FIRST:
  - If you need additional help with this template, please refer to https://www.home-assistant.io/help/reporting_issues/
  - Make sure you are running the latest version of Home Assistant before reporting an issue: https://github.com/home-assistant/core/releases
  - Do not report issues for integrations if you are using custom components or integrations.
  - Provide as many details as possible. Paste logs, configuration samples and code into the backticks.
  DO NOT DELETE ANY TEXT from this template! Otherwise, your issue may be closed without comment.
-->
## The problem
<!-- 
  Describe the issue you are experiencing here to communicate to the
  maintainers. Tell us what you were trying to do and what happened.
-->
`2020-09-13T21:44:53.3810115Z FAILED tests/components/recorder/test_util.py::test_last_run_was_recently_clean`

```
2020-09-13T21:44:53.1791416Z _______________________ test_last_run_was_recently_clean _______________________
2020-09-13T21:44:53.1793064Z [gw1] linux -- Python 3.8.2 /__w/core/core/venv/bin/python
2020-09-13T21:44:53.1793505Z 
2020-09-13T21:44:53.1794012Z hass_recorder = <function hass_recorder.<locals>.setup_recorder at 0x7f6927c66670>
2020-09-13T21:44:53.1794547Z 
2020-09-13T21:44:53.1799110Z     def test_last_run_was_recently_clean(hass_recorder):
2020-09-13T21:44:53.1799690Z         """Test we can check if the last recorder run was recently clean."""
2020-09-13T21:44:53.1800345Z         hass = hass_recorder()
2020-09-13T21:44:53.1803509Z     
2020-09-13T21:44:53.1803989Z         cursor = hass.data[DATA_INSTANCE].engine.raw_connection().cursor()
2020-09-13T21:44:53.1804601Z     
2020-09-13T21:44:53.1808492Z         assert util.last_run_was_recently_clean(cursor) is False
2020-09-13T21:44:53.1808895Z     
2020-09-13T21:44:53.1809240Z         hass.data[DATA_INSTANCE]._close_run()
2020-09-13T21:44:53.1812553Z         wait_recording_done(hass)
2020-09-13T21:44:53.1812911Z     
2020-09-13T21:44:53.1813350Z >       assert util.last_run_was_recently_clean(cursor) is True
2020-09-13T21:44:53.1817502Z E       assert False is True
2020-09-13T21:44:53.1820834Z E        +  where False = <function last_run_was_recently_clean at 0x7f693991f430>(<sqlite3.Cursor object at 0x7f692b371260>)
2020-09-13T21:44:53.1821818Z E        +    where <function last_run_was_recently_clean at 0x7f693991f430> = util.last_run_was_recently_clean
2020-09-13T21:44:53.1825216Z 
2020-09-13T21:44:53.1825758Z tests/components/recorder/test_util.py:156: AssertionError

```

--- a/homeassistant/components/recorder/__init__.py
+++ b/homeassistant/components/recorder/__init__.py
@@ -196,6 +196,12 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
 PurgeTask = namedtuple("PurgeTask", ["keep_days", "repack"])
 
 
+class WaitTask:
+    """An object to insert into the recorder queue to tell it set the _queue_watch event."""
+
+    pass
+
+
 class Recorder(threading.Thread):
     """A threaded recorder class."""
 
@@ -226,6 +232,7 @@ class Recorder(threading.Thread):
         self.db_retry_wait = db_retry_wait
         self.db_integrity_check = db_integrity_check
         self.async_db_ready = asyncio.Future()
+        self._queue_watch = threading.Event()
         self.engine: Any = None
         self.run_info: Any = None
 
@@ -352,6 +359,9 @@ class Recorder(threading.Thread):
                 # Schedule a new purge task if this one didn't finish
                 if not purge.purge_old_data(self, event.keep_days, event.repack):
                     self.queue.put(PurgeTask(event.keep_days, event.repack))
+                continue
+            if isinstance(event, WaitTask):
+                self._queue_watch.set()
                 continue
             if event.event_type == EVENT_TIME_CHANGED:
                 self._keepalive_count += 1
@@ -506,8 +516,9 @@ class Recorder(threading.Thread):
         after calling this to ensure the data
         is in the database.
         """
-        while not self.queue.empty():
-            time.sleep(0.025)
+        self._queue_watch.clear()
+        self.queue.put(WaitTask())
+        self._queue_watch.wait()
 
     def _setup_connection(self):
         """Ensure database is ready to fly."""

--- a/homeassistant/components/recorder/__init__.py
+++ b/homeassistant/components/recorder/__init__.py
@@ -199,8 +199,6 @@ PurgeTask = namedtuple("PurgeTask", ["keep_days", "repack"])
 class WaitTask:
     """An object to insert into the recorder queue to tell it set the _queue_watch event."""
 
-    pass
-
 
 class Recorder(threading.Thread):
     """A threaded recorder class."""

